pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java

```diff
@@ -103,18 +105,44 @@ public void join() throws InterruptedException {
     }
 
     private void startFunction(FunctionRuntimeInfo functionRuntimeInfo) throws Exception {
-        Function.Instance instance = functionRuntimeInfo.getFunctionInstance();
-        FunctionMetaData functionMetaData = instance.getFunctionMetaData();
+        FunctionMetaData functionMetaData = functionRuntimeInfo.getFunctionInstance().getFunctionMetaData();
+        int instanceId = functionRuntimeInfo.getFunctionInstance().getInstanceId();
         log.info("Starting function {} - {} ...",
-                functionMetaData.getFunctionDetails().getName(), instance.getInstanceId());
+                functionMetaData.getFunctionDetails().getName(), instanceId);
+        File pkgFile = null;
+        
+        String pkgLocation = functionMetaData.getPackageLocation().getPackagePath();
+        boolean isPkgUrlProvided = isFunctionPackageUrlSupported(pkgLocation);
+        
+        if(isPkgUrlProvided && pkgLocation.startsWith(Utils.FILE)) {
+            pkgFile = new File(pkgLocation);
+        } else {
+            downloadFile(pkgFile, isPkgUrlProvided, functionMetaData, instanceId);
```

Comment:
Is it retried when downloadFile comes back with exception?

Reply:
Thats good. How do we identify cases where it fails permanently(e.g: file not in the location)?